content/en/blog/_posts/2024-11-11-dynamic-kubernetes-api-server-for-cozystack/index.md

```diff
@@ -0,0 +1,496 @@
+that it serves certain APIs. Here's a simple example:
+
+```shell
+# kubectl get apiservices.apiregistration.k8s.io
+NAME                          	SERVICE                   	AVAILABLE   AGE
+v1alpha1.apps.cozystack.io    	cozy-system/cozystack-api 	True    	7h29m
+```
+
+As soon as the Kubernetes api-server receives requests for resources in the group
+`v1alpha1.apps.cozystack.io`, it redirects all those requests to our extension api-server,
+which can handle them based on the business logic we've built into it.
+
+
+## When to use the API Aggregation Layer
+
+The API Aggregation Layer helps solve several issues where the usual CRD mechanism might
+not enough. Let's break them down.
+
+### Imperative Logic and Subresources
+
+Besides regular resources, Kubernetes also has something called subresources.
+
+In Kubernetes, subresources are additional actions or operations you can perform on primary resources
+(like Pods, Deployments, Services) via the Kubernetes API. They provide interfaces to manage
+specific aspects of resources without affecting the entire object.
+
+A simple example is `status`, which is traditionally exposed as a separate subresource that you can
+access independently from the parent object. The `status` field isn't meant to be changed
+
+But beyond `/status`, Pods in Kubernetes also have subresources like `/exec`, `/portforward`, and
+`/log`. Interestingly, instead of the usual declarative resources in Kubernetes, these represent
+endpoints for imperative operations like viewing logs, proxying connections, executing commands in
+a running container, and so on.
+
+To support such imperative commands on your own API, you need implement an extension API and an
+extension API server. Here are some well-known examples:
+
+- **KubeVirt**: An add-on for Kubernetes that extends its API capabilities to run traditional virtual machines.
+  The extension api-server created as part of KubeVirt handles subresources
+  like `/restart`, `/console`, and `/vnc` for virtual machines.
+- **Knative**: A Kubernetes add-on that extends its capabilities for serverless computing,
+  implementing the `/scale` subresource to set up autoscaling for its resource types.
+
+By the way, even though subresource logic in Kubernetes can be *imperative*, you can manage access
+to them *declaratively* using Kubernetes standard RBAC model.
+
+For example this way you can control access to the `/log` and `/exec` subresources of the Pod kind:
+```yaml
+kind: Role
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  namespace: default
+  name: pod-and-pod-logs-reader
+rules:
+- apiGroups: [""]
+  resources: ["pods", "pods/log"]
+  verbs: ["get", "list"]
+- apiGroups: [""]
+  resources: ["pods/exec"]
+  verbs: ["create"]
+```
+
+### You're not tied to use etcd
+
+Usually, the Kubernetes api-server uses etcd for its backend.
+However, implementing your own api-server doesn't lock you into using only etcd.
```

Comment:
However, implementing your own API server doesn't lock you into using only etcd.